internal/adc/translator/tcproute.go

```diff
@@ -0,0 +1,112 @@
+	gatewayv1 "sigs.k8s.io/gateway-api/apis/v1"
+	gatewayv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+func newDefaultUpstreamWithoutScheme() *adctypes.Upstream {
+	return &adctypes.Upstream{
+		Metadata: adctypes.Metadata{
+			Labels: map[string]string{
+				"managed-by": "apisix-ingress-controller",
+			},
+		},
+		Nodes: make(adctypes.UpstreamNodes, 0),
+	}
+}
+
+func (t *Translator) TranslateTCPRoute(tctx *provider.TranslateContext, tcpRoute *gatewayv1alpha2.TCPRoute) (*TranslateResult, error) {
+	result := &TranslateResult{}
+	rules := tcpRoute.Spec.Rules
+	labels := label.GenLabel(tcpRoute)
+	for ruleIndex, rule := range rules {
+		service := adctypes.NewDefaultService()
+		service.Labels = labels
+		service.Name = adctypes.ComposeServiceNameWithStream(tcpRoute.Namespace, tcpRoute.Name, fmt.Sprintf("%d", ruleIndex))
+		service.ID = id.GenID(service.Name)
+		var (
+			upstreams = make([]*adctypes.Upstream, 0)
+		)
+		for _, backend := range rule.BackendRefs {
+			if backend.Namespace == nil {
+				namespace := gatewayv1.Namespace(tcpRoute.Namespace)
+				backend.Namespace = &namespace
+			}
+			upstream := newDefaultUpstreamWithoutScheme()
+			upNodes, err := t.translateBackendRef(tctx, backend, DefaultEndpointFilter)
+			if err != nil {
+				continue
+			}
+			if len(upNodes) == 0 {
+				continue
+			}
+			t.AttachBackendTrafficPolicyToUpstream(backend, tctx.BackendTrafficPolicies, upstream)
+			upstream.Nodes = upNodes
+			var (
+				kind string
+				port int32
+			)
+			if backend.Kind == nil {
+				kind = KindService
+			} else {
+				kind = string(*backend.Kind)
+			}
+			if backend.Port != nil {
+				port = int32(*backend.Port)
+			}
+			namespace := string(*backend.Namespace)
+			name := string(backend.Name)
+			upstreamName := adctypes.ComposeUpstreamNameForBackendRef(kind, namespace, name, port)
+			upstream.Name = upstreamName
+			upstream.ID = id.GenID(upstreamName)
+			upstreams = append(upstreams, upstream)
+		}
+
+		if len(upstreams) == 0 {
+			// Create a default upstream if no valid backends
+			upstream := newDefaultUpstreamWithoutScheme()
+			service.Upstream = upstream
+		} else if len(upstreams) == 1 {
+			// Single backend - use directly as service upstream
+			service.Upstream = upstreams[0]
+			// remove the id and name of the service.upstream, adc schema does not need id and name for it
+			service.Upstream.ID = ""
+			service.Upstream.Name = ""
+		} else {
+			service.Upstreams = upstreams
+		}
+		streamRoute := adctypes.NewDefaultStreamRoute()
+		streamRouteName := adctypes.ComposeStreamRouteName(tcpRoute.Namespace, tcpRoute.Name, fmt.Sprintf("%d", ruleIndex))
+		streamRoute.Name = streamRouteName
+		streamRoute.ID = id.GenID(streamRouteName)
+		streamRoute.Labels = labels
+		// TODO: support remote_addr, server_adrr, sni, server_port
+		service.StreamRoutes = append(service.StreamRoutes, streamRoute)
+		result.Services = append(result.Services, service)
```

Comment:
We must implement traffic-split here as well. Currently, the gateway does not support it, which requires some additional work.

Reply:
fixed